File: programmers/java/src/level3/dp/IntTriangle.java
package level3.dp;

import java.util.Arrays;
// https://programmers.co.kr/learn/courses/30/lessons/43105?language=java
public class IntTriangle {
    public static void main(String[] args) {
        IntTriangle it = new IntTriangle();
//        String input = "[[7], [3, 8], [8, 1, 0], [2, 7, 4, 4], [4, 5, 2, 6, 5]]";
//        input = input.replaceAll("\\[", "{").replaceAll("]", "}");
//        System.out.println("input = " + input);
        System.out.println(it.solution(new int[][]{{7}, {3, 8}, {8, 1, 0}, {2, 7, 4, 4}, {4, 5, 2, 6, 5}}));
        System.out.println(it.solution(new int[][]{{0}}));

    }

    public int solution(int[][] triangle) {
        int[] prev = triangle[0];
        for (int i = 1; i < triangle.length; i++) {
            int[] t = triangle[i];
            for (int j = 0; j < t.length; j++) {
                if( j == 0) {
                    t[j] += prev[0];
                    continue;
                }
                if ( j == t.length -1 ) {
                    t[j] += prev[j-1];
                    continue;
                }
                t[j] += Math.max(prev[j-1], prev[j]);
            }
            prev = t;
        }
        return Arrays.stream(prev).max().getAsInt();
    }

}

File: programmers/python/level3/dp/int_triangle.py
# https://programmers.co.kr/learn/courses/30/lessons/43105
def solution(triangle):
    prev = triangle[0]
    for t in triangle[1:]:
        for i in range(len(t)):
            if i == 0:
                t[i] += prev[0]
                continue
            if i == len(t) - 1:
                t[i] += prev[i-1]
                continue
            t[i] += max(prev[i-1], prev[i])
        prev = t
    return max(prev)


print(solution([[0]]))
print(solution([[7], [3, 8], [8, 1, 0], [2, 7, 4, 4], [4, 5, 2, 6, 5]]))
